### PR TITLE
GTEST: Add common function to find transports

### PR DESCRIPTION
**Pull request by dmitrygladkov**
## What

This PR adds `has_transport` function for UCT and UCP to find the transports used

## Why ?

Simplifies code by adding a common function

## How ?

1. Add virtual `has_transport` to `test_base` class
2. Implement `has_transport` in `uct_test` class
3. Implement `has_transport` in `ucp_test` class

### CODE REVIEW

File: test/gtest/uct/uct_test.h [hunk 1]
```diff
@@ -186,6 +186,32 @@ class uct_test : public testing::TestWithParam<const resource*>,
...
+        uct_tl_names() {
+            ib_ud_tl_names.push_back("ud");
+            ib_ud_tl_names.push_back("ud_mlx5");
+
+            ib_rc_tl_names.push_back("rc");
+            ib_rc_tl_names.push_back("rc_mlx5");
+
+            ib_rc_dc_tl_names.push_back("dc_mlx5");
+        }
+        std::vector<std::string> get_ib_ud() const {
```
Comment (hoopoepg) [inline test/gtest/uct/uct_test.h]: return std::vector<std::string>& instead of copy of vector

File: test/gtest/uct/uct_test.h [hunk 2]
```diff
@@ -186,6 +186,32 @@ class uct_test : public testing::TestWithParam<const resource*>,
...
+            return ib_ud_tl_names;
+        }
+        std::vector<std::string> get_ib_rc() const {
```
Comment (hoopoepg) [inline test/gtest/uct/uct_test.h]: return std::vector<std::string>& instead of copy of vector

File: test/gtest/uct/uct_test.h [hunk 3]
```diff
@@ -186,6 +186,32 @@ class uct_test : public testing::TestWithParam<const resource*>,
...
+            return ib_rc_tl_names;
+        }
+        std::vector<std::string> get_ib_rc_dc() const {
```
Comment (hoopoepg) [inline test/gtest/uct/uct_test.h]: return std::vector<std::string>& instead of copy of vector

File: test/gtest/uct/uct_test.h (hunk 1 above)
Comment (dmitrygladkov) [reply test/gtest/uct/uct_test.h]: good catch!
fixed

File: test/gtest/uct/uct_test.h (hunk 2 above)
Comment (dmitrygladkov) [reply test/gtest/uct/uct_test.h]: fixed

File: test/gtest/uct/uct_test.h (hunk 3 above)
Comment (dmitrygladkov) [reply test/gtest/uct/uct_test.h]: fixed

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9835/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9836/ for details (Mellanox internal link).

**Comment by dmitrygladkov** [comment]
@yosefe this PR is related to your comment in #3423. could you review pls?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9841/ for details (Mellanox internal link).

File: test/gtest/common/test.h [hunk 4]
```diff
@@ -66,6 +66,8 @@ class test_base {
 
     virtual void test_body() = 0;
 
+    virtual bool has_transport(const std::string& tl_name) const;
```
Comment (yosefe) [inline test/gtest/common/test.h]: why is it needed?

File: test/gtest/uct/uct_test.h [hunk 5]
```diff
@@ -186,6 +186,35 @@ class uct_test : public testing::TestWithParam<const resource*>,
         uct_iov_t               m_iov;
     };
 
+    class uct_tl_names {
```
Comment (yosefe) [inline test/gtest/uct/uct_test.h]: why need nested class and member variables and add std::vector<string> overload for has_transport?
can just define has_ud(), has_rc(), has_rc_or_dc() in uct_test

Comment (dmitrygladkov) [reply test/gtest/uct/uct_test.h]: good point, it simplifies
done

File: test/gtest/common/test.h (hunk 4 above)
Comment (dmitrygladkov) [reply test/gtest/common/test.h]: yes, there is no need
removed

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/9847/ for details (Mellanox internal link).

**Pull request merged**